src/tiled/objectsdock.cpp

```diff
@@ -454,3 +509,44 @@ void ObjectsView::synchronizeSelectedItems()
                              QItemSelectionModel::Clear);
     mSynching = false;
 }
+
+ObjectsFilterModel::ObjectsFilterModel(QObject *parent)
+    : QSortFilterProxyModel(parent)
+{
+}
+
+bool ObjectsFilterModel::filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const
+{
+    // Show all if filter string is empty. Prevents hiding empty groups
+    if(filterRegExp().isEmpty())
+        return true;
+
+    QModelIndex index = sourceModel()->index(sourceRow, 0, sourceParent);
+
+    // sourceParent of a group has no model
+    if(!sourceParent.model())
+        return groupHasAnyMatchingObjects(index);
+    else
+        return objectContainsFilterString(index);
+}
+
+bool ObjectsFilterModel::groupHasAnyMatchingObjects(const QModelIndex index) const
+{
+    for (int i = 0; i < sourceModel()->rowCount(index); ++i) {
+        QModelIndex childIndex = sourceModel()->index(i, 0, index);
+        if (childIndex.isValid() && objectContainsFilterString(childIndex))
```

Comment:
The `childIndex.isValid()` should not be necessary since you're only requesting data from valid indexes.